### PR TITLE
Unify app editor options menus

### LINKED ISSUE #1214: Unify app options menus
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

I would expect that the options menu associated with the app should be the same on the Home page (containing the list of apps) and the editor page. This would also involve moving the "View DOM" option into the menu, possibly renamed as "Export" (linked to https://github.com/mui/mui-toolpad/issues/1040)

### Examples 🌈

Screenshot 2022-10-26 at 12 25 48 PM

Screenshot 2022-10-26 at 12 28 34 PM

### Motivation 🔦

_No response_

### PR DESCRIPTION
**Pull request by bharatkashyap**
- Closes #1214 

- Visual demonstration 

Screenshot 2022-11-05 at 3 06 16 PM

Screenshot 2022-11-05 at 3 06 34 PM

- Please follow the [the contributing guide on how to create great pull requests](https://github.com/mui/mui-toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/toolpad-app/src/server/data.ts [hunk 1]
```diff
@@ -529,14 +529,19 @@ export async function loadRenderTree(appId: string, version: VersionOrPreview =
   return appDom.createRenderTree(await loadDom(appId, version));
 }
 
-export async function duplicateApp(id: string, name: string): Promise<AppMeta> {
+export async function duplicateApp(
+  id: string,
+  name: string,
+  suppliedName?: boolean,
```
Comment (Janpot) [inline packages/toolpad-app/src/server/data.ts:535]: I don't understand this parameter.

**Comment by prakhargupta1** [comment]
I have checked the UX of all menu options, and it looks good to me.

File: packages/toolpad-app/src/toolpad/AppEditor/PagePanel.tsx [hunk 2]
```diff
@@ -95,16 +50,33 @@ export default function PagePanel({ appId, className, sx }: ComponentPanelProps)
...
+            app={app}
+            onRename={handleRename}
+            onDuplicate={() => setDuplicateApp(app)}
+            onDelete={onDelete}
+            dom={dom}
+          />
+        ) : null}
       </Box>
       <Divider />
+      <AppDeleteDialog app={deletedApp} onClose={() => setDeletedApp(null)} redirectOnDelete />
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/PagePanel.tsx:74]: Would it make sense to include this in the menu as well?

File: packages/toolpad-app/src/server/data.ts (hunk 1 above)
Comment (bharatkashyap) [reply packages/toolpad-app/src/server/data.ts:535]: This is essentially to allow reusing the function in two scenarios:

1. When the user doesn't supply a name when duplicating (duplicating in the list view)
2. When we have a user supplied name while duplicating the app (duplicating in the editor view)

Comment (bharatkashyap) [reply packages/toolpad-app/src/server/data.ts:535]: On second thought, we can unify both these experiences and then get rid of the need of having to do this.

File: packages/toolpad-app/src/toolpad/AppEditor/PagePanel.tsx (hunk 2 above)
Comment (bharatkashyap) [reply packages/toolpad-app/src/toolpad/AppEditor/PagePanel.tsx:74]: Agreed, we can move both the `AppDeleteDialog` and `AppDuplicateDialog` inside the options menu and simplify the `duplicateApp` function as well

File: packages/toolpad-app/src/server/data.ts [hunk 3]
```diff
@@ -531,16 +531,26 @@ export async function loadRenderTree(appId: string, version: VersionOrPreview =
...
-    duplicateCount === 0 ? `${name} (copy)` : `${name} (copy ${duplicateCount + 1})`;
-  const newApp = await createApp(duplicateName, {
-    from: {
-      kind: 'dom',
-      dom,
-    },
-  });
-  return newApp;
+  try {
+    const newApp = await createApp(name, {
```
Comment (Janpot) [inline packages/toolpad-app/src/server/data.ts:535]: Let's extract and reuse the options?
```tsx
  const appInit: CreateAppOptions = {
    from: {
      kind: 'dom',
      dom,
    },
  };

    // ...
    const newApp = await createApp(name, appInit);
```

File: packages/toolpad-app/src/server/data.ts [hunk 4]
```diff
@@ -531,16 +531,26 @@ export async function loadRenderTree(appId: string, version: VersionOrPreview =
...
+      from: {
+        kind: 'dom',
+        dom,
+      },
+    });
+    return newApp;
+  } catch (error) {
+    const duplicateCount = await prismaClient.app.count({
```
Comment (Janpot) [inline packages/toolpad-app/src/server/data.ts:543]: This needs to only run when the error is the unique constraint of the name column. All other errors need to be rethrown. We wouldn't want this to run when something like a network error has happened

File: packages/toolpad-app/src/runtime/evalJsBindings.ts [hunk 5]
```diff
@@ -2,8 +2,7 @@ import { set } from 'lodash-es';
 import evalExpression from '../utils/evalExpression';
 import { mapValues } from '../utils/collections';
 import { errorFrom } from '../utils/errors';
-
-const TOOLPAD_LOADING_MARKER = '__TOOLPAD_LOADING_MARKER__';
```
Comment (Janpot) [inline packages/toolpad-app/src/runtime/evalJsBindings.ts:6]: This constant is not used outside of this file, let's keep it local?

File: packages/toolpad-app/src/server/data.ts [hunk 6]
```diff
@@ -231,7 +232,9 @@ export async function createApp(name: string, opts: CreateAppOptions = {}): Prom
   }
 
   if (await prismaClient.app.findUnique({ where: { name: appName } })) {
-    throw new Error(`An app named "${name}" already exists.`);
+    throw new Error(`An app named "${name}" already exists.`, {
+      cause: TOOLPAD_UNIQUE_CONSTRAINT_VIOLATION,
```
Comment (Janpot) [inline packages/toolpad-app/src/server/data.ts:236]: * We should only use `Error` instances in `cause`.
* Add a `code` property you import from `./src/errorCodes.tsx`:
    ```tsx
    export const ERR_APP_EXISTS = `TOOLPAD_ERR_APP_EXISTS`
    ```
    See also https://github.com/mui/mui-toolpad/pull/1272#issuecomment-1312401592

File: packages/toolpad-app/src/server/data.ts [hunk 7]
```diff
@@ -531,16 +534,25 @@ export async function loadRenderTree(appId: string, version: VersionOrPreview =
...
       dom,
     },
-  });
-  return newApp;
+  };
+  try {
+    const newApp = await createApp(name, appFromDom);
+    return newApp;
+  } catch (error) {
+    if (error instanceof Error && error.cause === TOOLPAD_UNIQUE_CONSTRAINT_VIOLATION) {
```
Comment (Janpot) [inline packages/toolpad-app/src/server/data.ts:547]: `TOOLPAD_UNIQUE_CONSTRAINT_VIOLATION` is an implementation detail. If we're adding our own error codes, let's do `ERR_APP_EXISTS`

File: packages/toolpad-app/src/constants.ts [hunk 8]
```diff
@@ -10,3 +10,5 @@ export const TOOLPAD_TARGET_PRO = 'PRO';
 export const DOCUMENTATION_URL = 'https://mui.com/toolpad/getting-started/overview/';
 export const DOCUMENTATION_INSTALLATION_URL =
   'https://mui.com/toolpad/getting-started/installation/';
+
+export const TOOLPAD_UNIQUE_CONSTRAINT_VIOLATION = 'UNIQUE_CONSTRAINT_VIOLATION';
```
Comment (Janpot) [inline packages/toolpad-app/src/constants.ts:14]:
```suggestion

```

File: packages/toolpad-app/src/errorCodes.ts [hunk 9]
```diff
@@ -1,2 +1,2 @@
-export const ERR_APP_NAME_EXISTS = 'TOOLPAD_APP_NAME_EXISTS';
+export const ERR_APP_EXISTS = 'TOOLPAD_APP_NAME_EXISTS';
```
Comment (Janpot) [inline packages/toolpad-app/src/errorCodes.ts:1]:
```suggestion
export const ERR_APP_EXISTS = 'TOOLPAD_APP_EXISTS';
```

**Review by Janpot — APPROVED** [review]

**Comment by Janpot** [comment]
One last nit in https://github.com/mui/mui-toolpad/pull/1286#discussion_r1031144001

**Pull request merged**